### DISCUSSION TITLE
QueriesDisabledError while logged into admin site

### DISCUSSION THREAD

**Discussion by silentjay** [dabapps/django-zen-queries#52] [Q&A]
When i'm logged into django admin site any access of the rest of the site results in a QueriesDisabledError. Is there a work around for this instead of logging out the admin site?

```
QueriesDisabledError at /events/search.html
SELECT "django_session"."session_key", "django_session"."session_data", 
"django_session"."expire_date" FROM "django_session" WHERE ("django_session"."expire_date" > %s 
AND "django_session"."session_key" = %s) LIMIT 21
```

**Comment by j4mie** [comment]
Hi,

I guess you're building a Django site that mostly works without the user being logged in, but when a user is logged in (in this case, as it happens, via the admin) you're getting errors?

I expect somewhere in your template you'll be accessing a property like `request.user.is_authenticated` which is causing the query in the session table. Without seeing your template or exactly how you're using zen-queries I can't help further, but the solution will probably be careful usage of `{% queries_dangerously_enabled %}` in your template around the part causing the query.

**Comment by silentjay** [comment]
Thanks for the response @j4mie 

I haven't accessed `request.user.is_authenticated` in my templates or views. However your comment got me thinking that a 3rd party library must be, dug a bit deeper and it seems that it's django debug toolbar causing the issue (although only on certain pages).

**Comment by j4mie** [comment]
Oh that's interesting. If I remember correctly, the debug toolbar injects itself into the response with a middleware rather than requiring an explicit template tag? That's a pain because it doesn't give you anywhere to "hook in" and wrap with `queries_dangerously_enabled`.

From a quick glance I can't see an obvious way to resolve this. Possibly a subclass of the djdt middleware or maybe even monkeypatching the `render_toolbar` method of the `DebugToolbar` class?

**Comment by silentjay** [comment]
Monkeypatching djdt is a bit above my paygrade, I'd say i'm only an advanced beginner python/django dev. 

A short term workaround I was trying to implement was to just not display the djdt toolbar when i'm logged in. However the act of just calling the `SHOW_TOOLBAR_CALLBACK` makes the zen-queries error disappear, still display the toolbar and I can stay logged in. Strange?

```py
settings.py

def show_toolbar(request):
    return request.user.username == "me"

DEBUG_TOOLBAR_CONFIG = {
    'SHOW_TOOLBAR_CALLBACK': 'core.settings.base.show_toolbar',
}
```

**Comment by j4mie** [comment]
Ah I think that makes sense - `request.user` is a lazy object, so touching it before the template is rendered means the query happens outside your `queries_disabled` block.

Out of interest, which approach to using `zen-queries` are you taking? Are you manually using the context manager, or the `render` shortcut, or something else? You might be able to also achieve this by using the "permissions gotcha" noted in the README: https://github.com/dabapps/django-zen-queries#permissions-gotcha

**Comment by silentjay** [comment]
Ah that makes perfect sense now with `request.user` being lazy. 

Currently using `zen-queries` with CBVs, setting the `response_class` attribute on the view to `zen_queries.TemplateResponse`.

Looked at the permissions gotcha you linked and tested `request.user.get_all_permissions()` in the view which also works. Although I think I'll stick with djdt callback workaround for now, as it means I won't have to remember to call `get_all_permissions()` in all my future views. 

Thanks again for helping get this working.
